nature/models.py

```diff
@@ -13,37 +13,87 @@
 from django.utils.translation import ugettext_lazy as _
 
 
-class ProtectionLevelEnabledQuerySet(models.QuerySet):
-    """Custom queryset to filter protection level enabled queryset based on user roles"""
+PROTECTION_LEVELS = {
+    'ADMIN_ONLY': 1,
+    'ADMIN_AND_STAFF': 2,
+    'PUBLIC': 3,
+}
+
+AUTHORIZATION_LEVELS = {
+    'ADMIN': 1,
+    'OFFICER': 2,
+    'PUBLIC': 3,
+    'OPEN_DATA': 4,
+}
+
+AUTHORIZATION_PROTECTION_MAP = {
+    AUTHORIZATION_LEVELS['ADMIN']: [
+        PROTECTION_LEVELS['ADMIN_ONLY'],
+        PROTECTION_LEVELS['ADMIN_AND_STAFF'],
+        PROTECTION_LEVELS['PUBLIC']
+    ],
+    AUTHORIZATION_LEVELS['OFFICER']: [
+        PROTECTION_LEVELS['ADMIN_AND_STAFF'],
+        PROTECTION_LEVELS['PUBLIC']
+    ],
+    AUTHORIZATION_LEVELS['PUBLIC']: [PROTECTION_LEVELS['PUBLIC']],
+    AUTHORIZATION_LEVELS['OPEN_DATA']: [PROTECTION_LEVELS['PUBLIC']],
+}
+
+
+class ProtectedQuerySet(models.QuerySet):
+    def __init__(self, model=None, *args, **kwargs):
+        self._model_fields = []
+        if model:
+            self._model_fields = [f.name for f in model._meta.get_fields()]
+        super().__init__(model=model, *args, **kwargs)
 
     def for_admin(self):
-        """ADMIN have access to all objects"""
-        return self
+        return self.filter_protected(AUTHORIZATION_LEVELS['ADMIN'])
 
     def for_admin_and_staff(self):
-        """ADMIN_AND_STAFF have access to objects of which the protection level is PUBLIC or ADMIN_AND_STAFF"""
-        return self.filter(protection_level__in=[ProtectionLevelMixin.PUBLIC, ProtectionLevelMixin.ADMIN_AND_STAFF])
+        return self.filter_protected(AUTHORIZATION_LEVELS['OFFICER'])
 
     def for_public(self):
-        """PUBLIC have access to objects of which the protection level is PUBLIC"""
-        return self.filter(protection_level=ProtectionLevelMixin.PUBLIC)
+        return self.filter_protected(AUTHORIZATION_LEVELS['PUBLIC'])
 
+    def open_data(self):
+        return self.filter_protected(AUTHORIZATION_LEVELS['OPEN_DATA'])
 
-class ProtectionLevelMixin(models.Model):
-    ADMIN_ONLY = 1
-    ADMIN_AND_STAFF = 2
-    PUBLIC = 3
+    def filter_protected(self, auth_level=AUTHORIZATION_LEVELS['OPEN_DATA']):
```

Comment:
I refactored the code once more and split it up into two functions. Now it should at least be somewhat more readable. I also added some function docstrings to make it clear what the functions does.

The nice thing is that this code keeps the same "API" (same functions in this case) as the previous solution. Which mean that even if I did not write any new tests, the old ones still work and passes. All of the different scenarios (admin/admin or staff/public/open) is already tested, at least with the simple tests that was there before. However, I do agree that more tests should be added, so that is what I did :)